src/pudl/transform/ferc1.py

```diff
@@ -1021,8 +1084,304 @@ def process_xbrl_metadata(self, xbrl_metadata_json) -> pd.DataFrame:
+
+        This method attempts to use the table params to translate a column name.
+
+        Note: Instead of doing this for each individual column name, we could compile a
+        rename dict for the whole table with a similar processand then apply it for each
+        group of columns instead of running through this full process every time. If
+        this took longer than...  ~5 ms on a single table w/ lots of calcs this would
+        probably be worth it for simplicity.
+        """
+        rename_dicts = [
+            self.params.rename_columns_ferc1.duration_xbrl,
+            self.params.rename_columns_ferc1.instant_xbrl,
+            self.params.rename_columns_ferc1.xbrl,
+        ]
+        col_name_new = col_name_xbrl
+        for rename_stage in rename_dicts:
+            col_name_new = str(rename_stage.columns.get(col_name_new, col_name_new))
+        # the values in wide_to_tidy are found at the end of the column names and
+        # extracted, so we need to remove all of the suffixes.
+        wide_to_tidy_value_types = []
+        for wide_to_tidy_stage in json.loads(self.params.wide_to_tidy.json()).values():
+            if not isinstance(wide_to_tidy_stage, list):
+                wide_to_tidy_stage = [wide_to_tidy_stage]
+            for wide_to_tidy in wide_to_tidy_stage:
+                if wide_to_tidy["value_types"]:
+                    wide_to_tidy_value_types.append(wide_to_tidy["value_types"])
+        wide_to_tidy_value_types = pudl.helpers.dedupe_n_flatten_list_of_lists(
+            wide_to_tidy_value_types
         )
 
+        for value_type in wide_to_tidy_value_types:
+            if col_name_new.endswith(f"_{value_type}"):
+                col_name_new = re.sub(f"_{value_type}$", "", col_name_new)
+            # col_name_new = col_name_new.replace(f"_{value_type}$", "")
+
+        if self.params.unstack_balances_to_report_year_instant_xbrl:
+            # TODO: do something...? add starting_balance & ending_balance suffixes?
+            pass
+        if self.params.convert_units:
+            # TODO: use from_unit -> to_unit map. but none of the $$ tables have this rn.
+            pass
+        return col_name_new
+
+    def remove_duplicated_components(self, tbl_meta: pd.DataFrame):
+        """If a calculation contains the same components >1x, remove duplicates."""
+        # reset the index bc we'll use it to compile a new series.
+        tbl_meta = tbl_meta.reset_index(drop=True)
+        new_calcs = pd.Series(dtype=pd.StringDtype())
+        for calc, index in zip(tbl_meta.calculations, tbl_meta.index):
+            calc = literal_eval(calc)
+            new_calc = [i for n, i in enumerate(calc) if i not in calc[n + 1 :]]
+            if new_calc != calc:
+                logger.info(
+                    f"Dropping duplicated components from calculation in {self.table_id.value}"
+                )
+            new_calcs.loc[index] = str(calc)
+        tbl_meta["calculations"] = new_calcs
```

Comment:
i replicated the json-based loopy-ness for the df version just for the most direct translation, but there is probably a cleaner/faster/less brittle way to do this. tbqh speed probably doesn't really matter bc there are not a lot of calcs but still